README.md
# Application Localizer Visual Studio Code Extension

Application Localizer Visual Studio Code Extension that helps with localizing applications.

[Pseudo Locale Generator Web Site](https://ppatotski.github.io/app-localizer/)

> Utilizes [app-localizer](https://github.com/ppatotski/app-localizer)
>
> [![NPM Package](https://img.shields.io/npm/v/app-localizer.svg)](https://www.npmjs.com/package/app-localizer)

## Features

* Locale validator
	- Check for missing labels
	- Multi-file locale support (see [example](#multi-file-locale-example) below)
		* Polymer file structure
		* Angular flat file structure
* Pseudo localizer
    - Accents on letters
    - Longer sentence
    - Longer word
    - Right-to-Left
    - Enclose in exclamations
    - Enclose in brackets
	- Support [ICU Message syntax](https://formatjs.io/guides/message-syntax/)
* Create workspace and global settings file (`Ctrl+Shift+P` or `Cmd+Shift+P` type `Applocalizer: Settings`)
	- Apply settings immediately after saving changes in `./.vscode/applocalizer.json` file
	- Support comments in JSON as an extension to JSON specification

## Get Started

[![Get Started](getstarted.gif)](getstarted.gif)

## Demo

[![Demo](demo.gif)](demo.gif)

## Settings Reference

Locale validator options
* **filePathPattern** Locale files path (supports node glob pattern).
* **multiFile** Each locale is in separate file in the same folder.
* **fileStructure** Structure of locale file content (`polymer` or `angular.flat` file structure).

Pseudo locale generator options
* **expander** Sentence expand factor 0.3 = 30%.
* **wordexpander** Word expand factor 0.5 = 50%.
* **brackets** Enclose sentence in brackets.
* **exclamations** Enclose sentence in exclamations.
* **accents** Convert letter to its accent version.
* **rightToLeft** RTL writing systems.
* **forceException** Force throwing syntax exception if any.
* **escapeCharacters** Escape special characters (`false` for json content)

## Config file example (`./.vscode/applocalizer.json` or `~/applocalizer.json`)
```json
{
	"validator": {
		"multiFile": false,
		"filePathPattern": "**/locales.json",
		"fileStructure": "polymer"
	},
	"pseudoLocale": {
		"expander": 0.3,
		"exclamations": true,
		"brackets": true,
		"accents": true,
		"rightToLeft": false,
		"wordexpander": 0.5,
		"forceException": false,
		"escapeCharacters": false
	}
}
```

## Single-File locale example `/locales/locale.json`
```json
{
	"en-us": {
		"label3": "blah3 {token}",
		"label1": "blah1",
		"label2": "blah2",
		"label4": "blah4"
	},
	"de-de": {
		"label3": "blah3 {token}",
		"label1": "blah1",
		"label2": "blah2",
		"label4": "blah4"
	}
}
```

## Multi-File locale example

### file with en-us locale `/locales/en-us.locale.json` (polymer file structure)
```json
{
	"en-us": {
		"label3": "blah3 {token}",
		"label1": "blah1",
		"label2": "blah2",
		"label4": "blah4"
	}
}
```

### file with fr locale `/locales/fr.locale.json` (angular flat file structure)
```json
{
	"label1": "blah1 {{token}}",
	"label2": "blah2",
	"label5": "blah3",
	"label3": "blah4"
}
```

## Change Log

[Change Log](CHANGELOG.md)

## License

[MIT](LICENSE.md)


CHANGELOG.md
# Change Log

## 0.3.11 (March 9, 2018)
- Issue: angular.flat locale handling issue fix

## 0.3.10 (September 11, 2017)
- Improvement: Added global settings support

## 0.3.9 (Aug 3, 2017)
- Improvement: Added json mode option that affects escaped characters handling
- Improvement: Updated dependency

## 0.3.8 (July 21, 2017)
- Improvement: Support comments in json configuration file
- Improvement: Support escaped characters

## 0.3.7 (July 14, 2017)
- Improvement: Update extension metadata and dependency

## 0.3.6 (June 5, 2017)
- Improvement: Update documentation and dependency

## 0.3.5 (May 11, 2017)
- Issue: generate unnecessary `null` in `{pctBlack, number, null}` for pseudo text.

## 0.3.4 (May 9, 2017)
- Improvement: Support ICU Message syntax

## 0.3.3 (May 5, 2017)
- Issue: Display `Applocalizer can only be enabled if VS Code is opened on a workspace folder` command message

## 0.3.2 (May 4, 2017)
- Bug: fixed several token handling issues for pseudo generator

## 0.3.1 (May 1, 2017)
- Improvement: Apply changes in `applocalizer.json` immediately

## 0.2.7 (April 25, 2017)
- Improvement: Updated improved version of app-localizer npm package

## 0.2.6 (April 24, 2017)
- Improvement: Externalizing validation logic into npm package

## 0.2.5 (April 21, 2017)
- Improvement: Pseudo locale algorithm improvement

## 0.2.4 (April 20, 2017)
- Improvement: Pseudo locale wordexpander option support

## 0.2.3 (April 19, 2017)
- Improvement: Externalizing logic into npm package

## 0.2.2 (April 18, 2017)
- Improvement: Locale string token support

## 0.2.1 (April 17, 2017)
- Improvement: Angular flat file structure support

## 0.2.0 (April 16, 2017)
- Improvement: Multi-file locale support

## 0.1.2 (April 14, 2017)
- Improvement: `Enclose in exclamations` is configurable

## 0.1.1 (April 13, 2017)
- Improvement: file path as a pattern (`filePathMatch` is deprecated)

## 0.1.0 (April 12, 2017)
- Features: Locale validator

## 0.0.3 (April 11, 2017)
- Features: Pseudo localizer
- Features: Create default settings


extension.js
const vscode = require("vscode");
const localizer = require("app-localizer");
const fs = require("fs");
const path = require("path");
const strip = require("strip-json-comments");

const exampleJson = `{
	"validator": {
		"multiFile": false,
		"filePathPattern": "**/locales.json",
		"fileStructure": "polymer"
	},
	"pseudoLocale": {
		"expander": 0.3,
		"wordexpander": 0.5,
		"exclamations": true,
		"brackets": true,
		"accents": true,
		"rightToLeft": false,
		"escapeCharacters": false
	}
}`;

function activate(context) {
	try {
		let workspaceSettings = false;
		let settingsPath = path.join(
      process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE,
      "applocalizer.json"
    );
		let settings = undefined;
		const vsSettingsCommand = vscode.commands.registerCommand("extension.applocalizer.settings", () => {
			fs.stat(settingsPath, err => {
				if (!err && workspaceSettings) {
					vscode.workspace.openTextDocument(settingsPath).then(doc => {
						vscode.window.showTextDocument(doc);
					});
				} else {
					const options = [];
					if (vscode.workspace && vscode.workspace.rootPath) {
						options.push("Create Workspace Settings");
					}
					if (err) {
						options.push("Create Global Settings");
					} else {
						options.push("Open Global Settings");
					}
					options.push("Cancel");

					vscode.window.showQuickPick(options).then(option => {
						let pathToSettings = settingsPath;
						if (option === "Create Workspace Settings") {
							pathToSettings = path.join(
                        vscode.workspace.rootPath,
                        ".vscode",
                        "applocalizer.json"
                      );
						}
						if (option === "Create Workspace Settings" || option === "Create Global Settings") {
							vscode.workspace
								.openTextDocument(
                          vscode.Uri.parse(`untitled:${pathToSettings}`)
                        )
								.then(doc => {
									vscode.window.showTextDocument(doc).then(editor => {
										editor.edit(edit => {
											edit.insert(new vscode.Position(0, 0), exampleJson);
										});
									});
								});
						} else if (option === "Open Global Settings") {
							vscode.workspace.openTextDocument(pathToSettings).then(doc => {
								vscode.window.showTextDocument(doc);
							});
						}
					});
				}
			});
		});

		context.subscriptions.push(vsSettingsCommand);
		const vsPseudoReplaceCommand = vscode.commands.registerCommand("extension.applocalizer.pseudoReplace", () => {
			vscode.window.activeTextEditor.edit(edit => {
				if (
            !vscode.window.activeTextEditor.selection.isEmpty &&
            settings &&
            settings.pseudoLocale
          ) {
					  let text = vscode.window.activeTextEditor.document.getText(
              vscode.window.activeTextEditor.selection
            );
					if (!settings.pseudoLocale.escapeCharacters) {
						text = JSON.parse(`{ "text": "${text}" }`).text;
					}
              // Need to think about better way to deal with escapes in text
					const result = localizer.toPseudoText(text, settings.pseudoLocale);
					edit.replace(vscode.window.activeTextEditor.selection, result);
				}
			});
		});

		context.subscriptions.push(vsPseudoReplaceCommand);
		const collection = vscode.languages.createDiagnosticCollection(
      "app-localizer"
    );

		context.subscriptions.push(collection);
		const processValidationResult = function processValidationResult(
      result,
      text,
      document
    ) {
			const diagnostics = [];
			if (result) {
				Object.keys(result).forEach(localeName => {
					Object.keys(result[localeName]).forEach(label => {
						let pos;
						if (settings.validator.fileStructure === "angular.flat") {
							pos = text.search(`"${label}"`);
						} else {
							const localeLocation = text.search(`"${localeName}"`);
							const location = text
                .substring(localeLocation)
                .search(`"${label}"`);
							pos = localeLocation + location;
						}
						const position = vscode.window.activeTextEditor.document.positionAt(pos);
						const range = new vscode.Range(
              position,
              new vscode.Position(
                position.line,
                position.character + `"${label}"`.length
              )
            );

						diagnostics.push(
              new vscode.Diagnostic(
                range,
                `Label "${label}" is missing in "${result[localeName][label].join()}" locale(s)`,
                vscode.DiagnosticSeverity.Error
              )
            );
					});
				});
			}
			collection.set(document.uri, diagnostics);
		};

		const validate = function validate(document) {
			if (
        settings &&
        settings.validator &&
        settings.validator.filePathPattern &&
        vscode.languages.match(
          { pattern: settings.validator.filePathPattern },
          document
        )
      ) {
				const text = document.getText();
				let result = undefined;
				if (settings.validator.multiFile) {
					const folder = path.dirname(document.fileName);
					localizer.validateLocales(
            folder,
            settings.validator,
            JSON.parse(text),
            result => processValidationResult(result, text, document)
          );
				} else {
					result = localizer.validateLocalesContent(JSON.parse(text));
					processValidationResult(result, text, document);
				}
			}
		};

		vscode.window.onDidChangeActiveTextEditor(event =>
      validate(event.document)
    );

		vscode.workspace.onDidChangeTextDocument(event => validate(event.document));

		const invalidateSettings = function invalidateSettings() {
			fs.stat(settingsPath, err => {
				if (!err) {
					fs.readFile(settingsPath, (err, buffer) => {
						if (err) {
							console.error(err);
						} else {
							settings = JSON.parse(strip(buffer.toString()));
							collection.clear();
							if (vscode.window.activeTextEditor) {
								validate(vscode.window.activeTextEditor.document);
							}
						}
					});
				} else {
					settings = undefined;
				}
			});
		};

		vscode.workspace.onDidSaveTextDocument(doc => {
			const selector = [];
			selector.push({ pattern: settingsPath });
			if (vscode.workspace && vscode.workspace.rootPath) {
				selector.push({
					pattern: path.join(
            vscode.workspace.rootPath,
            ".vscode",
            "applocalizer.json"
          )
				});
			}
			if (vscode.languages.match(selector, doc)) {
				if (
          vscode.workspace &&
          vscode.workspace.rootPath &&
          doc.fileName ===
            path.join(vscode.workspace.rootPath, ".vscode", "applocalizer.json")
        ) {
					settingsPath = doc.fileName;
					workspaceSettings = true;
				}
				invalidateSettings();
			}
		});

		if (vscode.workspace && vscode.workspace.rootPath) {
			const workspacePath = path.join(
        vscode.workspace.rootPath,
        ".vscode",
        "applocalizer.json"
      );
			fs.stat(workspacePath, err => {
				if (!err) {
					settingsPath = workspacePath;
					workspaceSettings = true;
				}
				invalidateSettings();
			});
		} else {
			invalidateSettings();
		}
	} catch (err) {
		console.error(err);
	}
}

exports.activate = activate;

function deactivate() {}

exports.deactivate = deactivate;
